styx-scheduler-service/src/test/java/com/spotify/styx/docker/KubernetesDockerRunnerTest.java

```diff
@@ -270,6 +281,116 @@ public void shouldRunIfSecretExists() throws IOException, StateManager.IsClosed
+    when(secretList.getItems()).thenReturn(ImmutableList.of(secret));
+
+    // Verify that the secret is delete even if we get permission denied errors on deleting the keys
+    final GoogleJsonResponseException permissionDenied = new GoogleJsonResponseException(
+        new Builder(403, "Forbidden", new HttpHeaders()), new GoogleJsonError()
+        .set("status", "PERMISSION_DENIED"));
+    doThrow(permissionDenied).when(serviceAccountKeyManager).deleteKey(keyName(SERVICE_ACCOUNT, "json-key"));
+    doThrow(permissionDenied).when(serviceAccountKeyManager).deleteKey(keyName(SERVICE_ACCOUNT, "p12-key"));
+    kdr.cleanup();
+    verify(serviceAccountKeyManager).deleteKey(keyName(SERVICE_ACCOUNT, "json-key"));
+    verify(serviceAccountKeyManager).deleteKey(keyName(SERVICE_ACCOUNT, "p12-key"));
+    verify(secrets).delete(secret);
+  }
+
+  @Test
+  public void shouldHandleErrorsWhenDeletingServiceAccountKeys() throws Exception {
+    final Secret secret1 = fakeServiceAccountKeySecret(SERVICE_ACCOUNT, SECRET_EPOCH, "json-key-1", "p12-key-1");
+    final Secret secret2 = fakeServiceAccountKeySecret(SERVICE_ACCOUNT, SECRET_EPOCH, "json-key-2", "p12-key-2");
+    final Secret secret3 = fakeServiceAccountKeySecret(SERVICE_ACCOUNT, SECRET_EPOCH, "json-key-3", "p12-key-3");
+
+    when(podList.getItems()).thenReturn(ImmutableList.of());
+    when(k8sClient.secrets()).thenReturn(secrets);
+    when(secrets.list()).thenReturn(secretList);
+    when(secretList.getItems()).thenReturn(ImmutableList.of(secret1, secret2, secret3));
+
+    when(secrets.delete(secret1)).thenThrow(new KubernetesClientException("fail delete secret1"));
+    doThrow(new IOException("fail delete json-key-2")).when(serviceAccountKeyManager).deleteKey("json-key-2");
+    doThrow(new IOException("fail delete p12-key-2")).when(serviceAccountKeyManager).deleteKey("p12-key-2");
+
+    kdr.cleanup();
+
+    verify(serviceAccountKeyManager).deleteKey(keyName(SERVICE_ACCOUNT, "json-key-1"));
+    verify(serviceAccountKeyManager).deleteKey(keyName(SERVICE_ACCOUNT, "p12-key-1"));
+    verify(secrets).delete(secret1);
+
+    verify(serviceAccountKeyManager).deleteKey(keyName(SERVICE_ACCOUNT, "json-key-2"));
+    verify(serviceAccountKeyManager).deleteKey(keyName(SERVICE_ACCOUNT, "p12-key-2"));
+    verify(secrets).delete(secret2);
+
+    verify(serviceAccountKeyManager).deleteKey(keyName(SERVICE_ACCOUNT, "json-key-3"));
+    verify(serviceAccountKeyManager).deleteKey(keyName(SERVICE_ACCOUNT, "p12-key-3"));
+    verify(secrets).delete(secret3);
+  }
+
+  private static Secret fakeServiceAccountKeySecret(String serviceAccount, String epoch, String jsonKeyId, String p12KeyId) {
+    final String jsonKeyName = keyName(serviceAccount, jsonKeyId);
+    final String p12KeyName = keyName(serviceAccount, p12KeyId);
+
+    final ObjectMeta metadata = new ObjectMeta();
+    metadata.setName("styx-wf-sa-keys-" + epoch + "-" + Hashing.sha256().hashString(serviceAccount, UTF_8));
```

Comment:
Yeah, one purpose of the tests is to lock down behavior of code in order to prevent backwards incompatible behavior changes. E.g., if the tests would just invoke `buildSecretName` directly, then a change of the `styx-wf-sa-keys` constant in `KubernetesDockerRunner` would pass the tests a little bit too easily. This way I feel there's a greater chance that we realize that changing that constant would cause styx to leave secrets orphaned in k8s that we'd need to handle etc.